File: auth/src/main/java/com/sherpa/auth/security/JwtService.java
package com.sherpa.auth.security;

import java.util.Date;

import javax.crypto.SecretKey;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import com.sherpa.auth.security.jwt.AccessTokenPayload;

import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;

@Service
public class JwtService {

	private final SecretKey secretKey;

	@Value("${spring.application.name")
	private String issuer;

	@Value("${jwt.access-key-expiration-ms}")
	private long accessKeyExpirationInMs;

	public JwtService(@Value("${jwt.secret-key}") String secretKey) {
		this.secretKey = Keys.hmacShaKeyFor(Decoders.BASE64.decode(secretKey));
	}

	public String createAccessToken(AccessTokenPayload jwtPayload) {

		return Jwts.builder()
			.subject(jwtPayload.email())
			.issuer(issuer)
			.issuedAt(jwtPayload.issuedAt())
			.expiration(new Date(jwtPayload.issuedAt().getTime() + accessKeyExpirationInMs))
			.signWith(secretKey)
			.compact();
	}

}

File: gateway/src/main/kotlin/com/sherpa/gateway/domain/auth/jwt/JwtService.kt
package com.sherpa.gateway.domain.auth.jwt

import io.jsonwebtoken.Jwts
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.util.*
import javax.crypto.SecretKey

@Service
class JwtService(
    @Value("\${jwt.secret-key}")
    secretKeyBase64: String
) {
    private val secretKey: SecretKey = Keys.hmacShaKeyFor(Decoders.BASE64.decode(secretKeyBase64))

    @Value("\${spring.application.name}")
    private lateinit var issuer: String

    @Value("\${jwt.access-key-expiration-s}")
    private var accessKeyExpirationInS: Long = 0

    private val accessKeyExpirationInMs: Long
        get() = accessKeyExpirationInS * 100


    fun createAccessToken(jwtPayload: AccessTokenPayload): String {

        return Jwts.builder()
            .subject(jwtPayload.email)
            .issuer(issuer)
            .issuedAt(jwtPayload.issuedAt)
            .expiration(Date(jwtPayload.issuedAt.time + accessKeyExpirationInMs))
            .signWith(secretKey)
            .compact()
    }

}